### ISSUE TITLE
Default allow (blacklist) option

### ISSUE THREAD

**Issue by ashketchupandmustard** [julian-klode/dns66#75]
Please add a default allow (blacklist) option using addAllowedApplication for users who wish to block only a few specific apps.

**Comment by julian-klode** [comment]
I'm thinking about this. I was also thinking about disallowing system apps by default (with some exceptions), while keeping installed apps use the VPN - after all, pre-installed apps are less likely to have ads.

But it's arguably a bit hard to write. Imagine toggling the default with some stuff having custom values. What is the correct user experience here: Keep unchanged values at their default (that is, switch them over); or should I just take the inverse (that is, you change from allow to disallow, and I'd store all allowed apps in the config). They first loses state, which is a bit unfortunate (toggling the default twice does not get you back to the initial state).

**Comment by smichel17** [comment]
Is the ability to block all traffic to/from an app in the scope of DNS66?

**Comment by julian-klode** [comment]
Not right now, see #30. Disallowing here means allowing an app not to use the VPN, #30 deals with forcefully returning NXDOMAIN for specific apps

**Comment by smichel17** [comment]
The UX I would suggest here would be:

- Allow users to set each app as whitelisted (never block ads), blacklisted (always block ads) or default.
- Have a setting `Block ads by default` that is turned on by default.
  - If a user chooses white/blacklist, always obey that preference regardless of the default setting.
  - If a user chooses default, obviously you follow the setting
- Move this to its own subscreen, and make the `Whitelist apps` menu item open this subscreen instead of opening the dialogue
  - You can re-use the layout from the `Domain Filters` tab.

From the backend perspective, I'd probably store a boolean pref using the app name as the key, and I might use a different sharedPreferences than the default, to avoid namespace collisions (however unlikely) with your actual preferences.

**Comment by julian-klode** [comment]
Yes, storing the two states explictly solves the issue. Well, actually, with the firewall stuff from #30 we have more options (AFAICT: Allow, Apply host-based rules, Block requests - actually, once for each connectivity type). The default could then be toggled between Bypass and Host-based.

BTW, two minor points:

1. I'm not going to mention ads anywhere
2. all settings are stored in a single json file (which also allows you to sync them)

But I really thought more about doing #30 first, and then this one, as the rework for #30 basically allows this.

**Comment by julian-klode** [comment]
Well, and also a combined fourth state: Custom. So you can either select one type for all connections, or custom and configure stuff per connection type.

**Comment by smichel17** [comment]
The fourth state is confusing to me.

**Comment by julian-klode** [comment]
@smichel17  Think about it like this:

State: [Allow/Host-based/Block/Custom]

when you select custom, the following appears:

When on Wi-Fi: [Allow/Host-based/Block/Custom]
When on Cellular: [Allow/Host-based/Block/Custom]

Or maybe metered vs unmetered connections, not sure.

**Comment by smichel17** [comment]
What is custom?

**Comment by smichel17** [comment]
I would set that as two settings, both of which are always visible:

When on Wi-Fi: [Allow/Host-based/Block]
When on Cellular: [Allow/Host-based/Block/Same as Wi-Fi]

**Comment by smichel17** [comment]
Also, "Host-based" sounds a little awkward. Maybe "Filter" is a good one-word replacement?

**Comment by julian-klode** [comment]
Really the reason I want the global toggle is so I can have some state switching directly in the app list without having to go into the detail for the app. But I guess we can also do this with the two settings. Not sure how many settings there will be.

And I don't think "Filter" is a good alternative. We already have a "Hosts" tab where you can configure host-based rules, so host-based rules (or well, "Filter by Host"?) is a bit easier to understand than just filter (we can use one to three words in the spinner, I think)

**Comment by smichel17** [comment]
After a little thinking, I would implement this in a slightly different way than I suggest above. Instead, I would:

- Put the global default settings on the start/stop page (which I would rename).
  - For now, just shrink the power button, which doesn't need to be that big. In the long term I'd suggest moving the power button to the app bar or something like that, but that's not important for now. 
- Make the per-app rules page follow the layout of the other two much more closely.
  - Switch at the top for global enable/disable
  - FAB at the bottom for adding a rule for a new app
  - Only apps that actually have rules added are displayed (although, like the other screens you can disable these individually without removing them from the list.

Advantages over the previous design:

- Better separation of different parts of the mental model (per-app rules vs rules on different networks)
- Easy way to temporarily disable per-app rules without losing what you had
- Perhaps most importantly, no need to scroll through your entire list of installed apps just to see what rules you've set. 
- Closer to the current interface; less work for you!

**Comment by julian-klode** [comment]
> Put the global default settings on the start/stop page (which I would rename).

Which global default settings? Are you saying you want the per-connection-type defaults on the start page, and then only allow overriding the values per app? That probably makes sense.

> Make the per-app rules page follow the layout of the other two much more closely.

That is one of the possibilities. I already did stage 1 a few weeks ago - move the whitelist dialog into an apps tab (which only exists in master, and in the preview on XDA, not sure if you've seen it, screenshots are [post 303](https://forum.xda-developers.com/showpost.php?p=71204797&postcount=303)).

Stage 2 is to add other types of rules. Stage 3 would be to only show the currently configured apps, and have the floating action button. These are independent stages, though, so it might happen the other way around :). 

Showing only the configured apps means adding at least 2 new layout files, a new adapter, and a new activity - the current list of all apps becomes a *select an app* activity dialog (?), sans checkboxes - and a new list that just shows current apps replaces the current tab. It's a bit annoying work, but it's likely going to make the UI a lot better.

> Better separation of different parts of the mental model (per-app rules vs rules on different networks)

JFTR: I never really wanted multiple types of rules. It's mostly an implementation artefact: Whitelisting is done outside of our filtering, and thus needs to be applied first. Apart from whitelisting, everything follows a table driven (condition, rule) approach (or well, that's the theory).

Maybe we end up with a concept in the future that merges rules for apps together with rules for hosts and lets you construct more complex rules like "If * host is in host file ... and * we are on metered network then REJECT". I'd like that, but I also noticed that people already have issues with the current rules framework where order is relevant (the last action for a host applies) - a lot of people seem to prefer a whitelist/blacklist approach.

> Closer to the current interface; less work for you!

Don't assume that just because something looks similar it is easier to implement. It might be the other way around.
